db/snapshot/backup_task.cc

```diff
@@ -99,52 +111,132 @@ future<> backup_task_impl::upload_component(sstring name) {
     std::exception_ptr ex;
-    gate uploads;
-    auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+    auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+    std::vector<std::tuple<size_t, sstring>> backup_files;
+    std::optional<directory_entry> component_ent;
 
-    for (;;) {
-        std::optional<directory_entry> component_ent;
+    do {
         try {
             component_ent = co_await snapshot_dir_lister.get();
-        } catch (...) {
-            if (!ex) {
-                ex = std::current_exception();
-                break;
+            if (component_ent) {
+                auto component_name = snapshot_dir / component_ent->name;
+                auto size = co_await file_size(component_name.native());
+                backup_files.emplace_back(size, std::move(component_ent->name));
             }
-        }
-        if (!component_ent.has_value()) {
+        } catch (...) {
+            ex = std::current_exception();
             break;
         }
-        auto gh = uploads.hold();
+    } while (component_ent);
+    co_await snapshot_dir_lister.close();
+    if (ex)
+        co_await coroutine::return_exception_ptr(std::move(ex));
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    co_return greedy_partitioning(std::move(backup_files), smp::count);
+}
+} // namespace
 
-        std::ignore = upload_component(component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
+    auto backup_shards = co_await get_backup_files(_snapshot_dir);
+
+    if (snap_log.is_enabled(log_level::debug))
+        snap_log.debug("{}", print_results(backup_shards));
+
+    std::vector<abort_source> as_per_shard(smp::count);
+    auto s = _as.subscribe([&]() noexcept {
+        try {
+            std::ignore = smp::invoke_on_all([&as_per_shard, ex = _as.abort_requested_exception_ptr()] {
+                as_per_shard[this_shard_id()].request_abort_ex(ex);
```

Comment:
Ignoring a future is a sure way to get use-after-free. It requires a comment explaining why it's safe (when it is), or not to do it (when it's not, like here).

Reply:
`co_await`ing now